router/sources/router.move

```diff
@@ -0,0 +1,928 @@
+        };
+        if (mode == MODE_V1_AND_V2) {
+            v2_domains::set_subdomain_transferability_as_domain_owner(
+                &get_router_signer(),
+                user,
+                domain_name,
+                subdomain_name,
+                transferrable
+            );
+        };
+
+        set_primary_name_when_register(
+            user,
+            target_addr,
+            to_addr,
+            domain_name,
+            option::some(subdomain_name),
+        );
+    }
+
+    // ==== MIGRATION ====
+
+    public entry fun migrate_name(
+        user: &signer,
+        domain_name: String,
+        subdomain_name: Option<String>,
+    ) acquires RouterConfig {
+        let mode = get_mode();
+        if (mode == MODE_V1) {
+            abort error::not_implemented(ENOT_IMPLEMENTED_IN_MODE)
+        } else if (mode == MODE_V1_AND_V2) {
+            let user_addr = address_of(user);
+            // Check name is not already migrated
+            assert!(
+                !exists_in_v2(domain_name, subdomain_name),
+                error::invalid_state(ENAME_ALREADY_MIGRATED)
+            );
+
+            let (is_v1_owner, _token_id) = domains::is_token_owner(
+                user_addr,
+                subdomain_name,
+                domain_name,
+            );
+            assert!(is_v1_owner, error::permission_denied(ENOT_NAME_OWNER));
+            assert!(
+                !domains::name_is_expired_past_grace(subdomain_name, domain_name),
+                error::invalid_state(EMIGRATION_ALREADY_EXPIRED)
+            );
+
+            // Check primary name status
+            let maybe_primary_name = domains::get_reverse_lookup(user_addr);
+            let is_primary_name = if (option::is_some(&maybe_primary_name)) {
+                let (primary_subdomain_name, primary_domain_name) = domains::get_name_record_key_v1_props(
+                    &option::extract(&mut maybe_primary_name)
+                );
+                subdomain_name == primary_subdomain_name && domain_name == primary_domain_name
+            } else {
+                false
+            };
+
+            // Get the v1 token info
+            let (
+                _property_version,
+                expiration_time_sec,
+                target_addr
+            ) = domains::get_name_record_v1_props_for_name(
+                subdomain_name,
+                domain_name,
+            );
+            let tokendata_id = aptos_names::token_helper::build_tokendata_id(
+                aptos_names::token_helper::get_token_signer_address(),
+                subdomain_name,
+                domain_name,
+            );
+            let token_id = aptos_names::token_helper::latest_token_id(&tokendata_id);
+
+            // Domain must migrate before subdomain, throw error if this is a subdomain but domain has not been migrated
+            if (option::is_some(&subdomain_name)) {
+                assert!(
```

Comment:
Do we actually need this check? Users can't register subdomains without the domain existing right?

Reply:
Is there not a check in the registration step already?